apps/workshop/projects/Perp88/helpers/calculation.ts

```diff
@@ -0,0 +1,105 @@
+import { Context } from '@defiyield/sandbox';
+import { BigNumber, ethers } from 'ethers';
+import {
+  CHAINLINK_ORACLE_ADDR,
+  COMPOSITION_TOKENS,
+  PLP_STAKING_ADDR,
+  PLP_STAKING_REVENUE_ADDR,
+  PLP_TOKEN_ADDR,
+  POOL_DIAMOND_ADDR,
+  ALL_TOKENS,
+} from './config';
+import GetterFacetAbi from '../abis/GetterFacet.json';
+import FeedableRewarderAbi from '../abis/FeedableRewarder.json';
+import PLPStakingAbi from '../abis/PLPStaking.json';
+import PoolOracleAbi from '../abis/PoolOracle.json';
+import ERC20Abi from '../abis/ERC20.json';
+import { e18, secondsInYear, Zero } from './constant';
+
+interface TokenPrice {
+  minPrice: BigNumber;
+  maxPrice: BigNumber;
+  avgPrice: BigNumber;
+  displayPrice: string;
+}
+
+const decimalPlaceDefault = 2;
+export async function calAPR(ctx: Context): Promise<string> {
+  const { ethcall, ethcallProvider, logger } = ctx;
+
+  const rewardContract = new ethcall.Contract(PLP_STAKING_REVENUE_ADDR, FeedableRewarderAbi);
+  const plpStakingContract = new ethcall.Contract(PLP_STAKING_ADDR, PLPStakingAbi);
+
+  const [rewardRate, totalShare]: BigNumber[] = await ethcallProvider.all([
+    rewardContract.rewardRate(),
+    plpStakingContract.calculateTotalShare(PLP_STAKING_REVENUE_ADDR),
+  ]);
+
+  const rewardRateE18 = rewardRate.mul(10 ** 12); //usdc is 1e6
+
+  const price = await getTokenOraclePrice(ctx, COMPOSITION_TOKENS.USDC);
+
+  const plpPrice = await getPLPPrice(ctx);
+
+  const secondsInYearBN = e18.mul(secondsInYear); //e18
+  const rewardRatePerYear = secondsInYearBN.mul(rewardRateE18).div(e18);
+  const rewardPricePerYear = rewardRatePerYear.mul(price.avgPrice).div(e18);
+  const totalRewardToken = totalShare.mul(plpPrice.avgPrice).div(e18);
+
+  if (totalRewardToken.isZero()) throw new Error(`_calAPR totalReward is 0`);
+
+  const apr = ethers.utils.formatEther(rewardPricePerYear.mul(e18).div(totalRewardToken).mul(100));
+
+  return apr;
+}
+
+// PRICES
+export async function getPLPPrice(ctx: Context): Promise<TokenPrice> {
```

Comment:
We added PLP asset tracking to our dashboard, so now you can delete this price logic from your PR